src/rpc/nfs_program.rs

```diff
@@ -0,0 +1,153 @@
+use std::collections::HashMap;
+use std::path::{PathBuf, Path};
+use std::fs::File;
+use std::os::unix::fs::MetadataExt;
+
+use tokio_util::udp::UdpFramed;
+use tokio::stream::StreamExt;
+use futures::{SinkExt};
+
+use crate::rpc::fs::get_fhandle;
+use crate::rpc::packets::{
+    *,
+    self as rpc_packages,
+    NfsLookupReply,
+    NfsFileAttributes,
+    NfsStatus,
+};
+use crate::rpc::codec::RpcBytesCodec;
+use crate::rpc::fs::read_file_range;
+
+pub struct RpcNfsProgramHandler {
+    path: PathBuf,
+    file_handlers: HashMap<u64, File>,
+}
+
+#[derive(Debug)]
+pub enum NfsProcedureError {
+    FileDoesNotExist,
+    StaleFileHandle,
+    NotImplemented,
+}
+
+impl From<std::io::Error> for NfsProcedureError {
+    fn from(_error: std::io::Error) -> NfsProcedureError {
+        NfsProcedureError::FileDoesNotExist
+    }
+}
+
+impl RpcNfsProgramHandler {
+    pub fn new() -> Self {
+        Self {
+            path: PathBuf::from("/"),
+            file_handlers: HashMap::new(),
+        }
+    }
+
+    fn reset_path(&mut self) {
+        self.path = PathBuf::from("/");
+    }
+
+    pub fn lookup(&mut self, lookup: &rpc_packages::NfsLookup) -> Result<NfsLookupReply, NfsProcedureError> {
+        let mut temp_path = self.path.clone();
+        temp_path.push(lookup.filename());
+
+        match std::fs::metadata(temp_path.as_path()) {
+            Ok(metadata) => {
+                self.path = temp_path;
+                let fwrapper = get_fhandle(self.path.as_path(), metadata.ino())?;
+
+                if metadata.is_file() {
+                    self.file_handlers.insert(fwrapper.inode, fwrapper.file);
```

Comment:
Just want to say that this implementation is the most naive thing I have ever written! This will eventually make the program crash with TooManyOpenFiles :-)